schema-engine/connectors/sql-schema-connector/src/sql_schema_differ.rs

```diff
@@ -538,6 +538,101 @@ fn is_prisma_implicit_m2m_fk(fk: ForeignKeyWalker<'_>) -> bool {
     table.column("A").is_some() && table.column("B").is_some()
 }
 
-fn all_match<T: PartialEq>(a: &mut dyn ExactSizeIterator<Item = T>, b: &mut dyn ExactSizeIterator<Item = T>) -> bool {
+fn all_match<T: PartialEq>(a: impl ExactSizeIterator<Item = T>, b: impl ExactSizeIterator<Item = T>) -> bool {
     a.len() == b.len() && a.zip(b).all(|(a, b)| a == b)
 }
+
+fn sort_migration_steps(steps: &mut Vec<SqlMigrationStep>, db: &DifferDatabase<'_>) {
+    // We can't merge these two steps into `sort_by` because sorting algorithms require total order
+    // relation, but the dependency between dropped unique index and creating a corresponding
+    // primary key is a preorder. Moreover, the binary relation defined as the composition of
+    // custom logic for `(SqlMigrationStep::DropIndex, SqlMigrationStep::AlterTable)` pairs and
+    // `a.cmp(b)` for everything else doesn't appear to be even transitive (due to the existing
+    // automatically derived total order relation between `SqlMigrationStep`s). If we need more
+    // complex ordering logic in the future, we should consider defining a partial order on
+    // `SqlMigrationStep` where only the pairs for which order actually matters are ordered,
+    // building a graph from the steps and topologically sorting it.
```

Comment:
Topological graph + explicit partial order would probably be the best course of action in the long term.
Currently, we need to manually be careful when defining the operations that impact the `schema.prisma` <-> `SQL` translation.